File: apps/inference/neuronpedia_inference/schemas/lens.py
"""Wire models for ``/v1/lens/*``.

Hand-written rather than generated, so the annotations stay loose (see the note in
``activation.py``). Only the request side lives here; the streamed NDJSON frames stay in
``endpoints/lens/prompt.py`` next to the generator that emits them, because they are not a
response body FastAPI can document.
"""

from enum import StrEnum

from neuronpedia_inference.schemas.common import BaseSchema


class LensType(StrEnum):
    """Which lens's readout direction to use."""

    LOGIT_LENS = "LOGIT_LENS"
    JACOBIAN_LENS = "JACOBIAN_LENS"


class LensChatMessage(BaseSchema):
    """One message of a chat-formatted lens prompt."""

    role: str
    content: str


class LensSteerToken(BaseSchema):
    """A single readout to steer on.

    ``token`` is the EXACT decoded token string (whitespace preserved, e.g.
    ``" cat"``) as it appeared in a read-out slice; the server resolves it back
    to a vocab id via a cached reverse-decode map. ``type`` selects which lens's
    readout direction to use: ``JACOBIAN_LENS`` uses the J-lens direction
    ``J_bar_l^T @ w_t`` at each fitted layer (the residual-space direction whose
    J-lens readout is this token), ``LOGIT_LENS`` uses the plain unembedding
    direction ``w_t``.
    """

    token: str
    type: LensType


class LensPromptRequest(BaseSchema):
    """Everything one lens run needs: what to read out, where, and how to intervene."""

    model: str
    # One or more lens types to compute. When both are requested, the model is
    # run only once (the residuals are shared), so adding LOGIT_LENS alongside
    # JACOBIAN_LENS is essentially free.
    type: list[LensType]
    # Provide exactly one of `prompt` (raw text) or `chat` (chat-formatted).
    prompt: str | None = None
    chat: list[LensChatMessage] | None = None
    top_n: int = 10
    # Layers to read out. Empty (default) = all available layers for the lens
    # type. The model's final layer is ALWAYS included (decoded directly as the
    # model's true output), regardless of this list.
    layers: list[int] = []
    max_seq_len: int | None = None
    prepend_bos: bool = True
    # Whether to enable "thinking" mode when applying a chat template (only
    # relevant for `chat` requests on models whose chat template supports it).
    enable_thinking: bool = False
    # Whether to preserve historical reasoning (`<think>`) blocks when applying a
    # chat template (only relevant for `chat` requests on models whose chat
    # template supports it, e.g. Qwen3.6). Keeping historical think blocks (the
    # default) stabilizes the chat-formatted token prefix across turns: without
    # it, templates strip prior `<think>` blocks from history, shifting token
    # positions every turn.
    preserve_thinking: bool = True
    # Stream results as NDJSON (one message per line). When false, the identical
    # path runs and all messages are buffered into a single JSON object.
    stream: bool = True
    # Sampling temperature for generated tokens. 0 = greedy (argmax).
    temperature: float = 1.0
    # Number of tokens to generate after the prompt. 0 = lens over the prompt
    # only (no generation).
    num_completion_tokens: int = 0
    # Token ids the client already has lens read-outs for (the previous
    # response's prompt + generated tokens, in order). The server computes the
    # longest common prefix with the freshly tokenized prompt and skips the
    # (expensive) per-layer read-out + emission for those positions, so a
    # follow-up turn only recomputes the new tokens. Position reuse is validated
    # by token id, so a divergent prefix simply reuses less (never wrong).
    cached_token_ids: list[int] = []
    # Exact input token ids to read out over, bypassing tokenization. When
    # provided (non-empty), `prompt`/`chat` are ignored, generation is disabled
    # (``num_completion_tokens`` is forced to 0), and the read-out runs over
    # exactly these ids. Used to faithfully reproduce a previously-computed run
    # (e.g. a shared jlens link) without depending on chat-template / tokenizer
    # drift — the lens read-out is a deterministic function of the token ids.
    input_token_ids: list[int] = []
    # Steering: readouts to additively inject (negatively, to suppress) into the
    # residual stream at every position, during prefill AND generation. Empty
    # (default) = no steering. When steering is active, prefix-reuse is disabled
    # (the cached read-outs from an unsteered run are no longer valid).
    steer_tokens: list[LensSteerToken] = []
    # Layers to inject the steering direction at. Empty = the read-out layers.
    steer_layers: list[int] = []
    # Signed steering strength as a fraction of each position's residual norm
    # (negative suppresses the readout). 0 = no steering.
    steer_strength: float = 0.0
    # When true, ABLATE the readout direction: project it out of the residual at
    # every steered layer/position (h <- h - (h.d_hat) d_hat) instead of
    # additively steering. Mutually exclusive with ``steer_strength`` (which is
    # ignored when ablating).
    steer_ablate: bool = False
    # SWAP: when set, replace the source readout (``steer_tokens[0]``) with this
    # target readout at every steered layer/position. The residual's projection
    # onto the source direction is removed and re-added (same magnitude) along
    # the target direction: ``h <- h - (h.s_hat) s_hat + (h.s_hat) t_hat``. This
    # is the causal "lens-vector swap" intervention and takes precedence over
    # ``steer_strength`` / ``steer_ablate`` when present. ``type`` should match
    # the source readout's lens type.
    swap_token: LensSteerToken | None = None
    # Whether to apply the steer/swap intervention to GENERATED tokens too. When
    # false (default), the intervention is applied only to the prompt positions
    # (prefill); generation then proceeds from the steered prompt context but the
    # newly generated positions are not themselves steered/swapped. When true,
    # the intervention is also applied at each generated position as it is
    # produced.
    steer_generated_tokens: bool = False
    # Whether to drop "non-word" tokens (punctuation / whitespace / symbol /
    # special tokens) from each position's per-layer read-out BEFORE selecting
    # the top-n, so the returned tokens are predominantly interesting word
    # tokens. The model's TRUE top-1 (output) token at each layer is always
    # preserved even when it is non-word. Probabilities are computed over the
    # FULL vocab (filtering only changes WHICH tokens are selected, not their
    # reported probabilities). Defaults to True.
    filter_non_word_tokens: bool = True
    # When true, if the server is already processing another request (the global
    # model lock is held), return HTTP 429 immediately instead of queueing and
    # waiting for the lock. This lets a client (e.g. the webapp) fail over to a
    # different inference server for this model. Defaults to False, preserving
    # the original behavior of waiting up to REQUEST_LOCK_TIMEOUT for the lock.
    fail_if_busy: bool = False

File: apps/webapp/lib/utils/lens.ts
import type { components } from '@/lib/api/inference';
// Shared types + constants for the streaming Jacobian/Logit lens endpoint
// (`POST /v1/lens/prompt` on the inference server). Pure types only — safe to
// import from both server and client code.
//
// The endpoint streams NDJSON (one JSON message per line): a single `meta`
// message, then one `token` message per sequence position, then a final
// `done` message (or an `error` message). When `stream` is false the same
// messages are buffered into a single `{ meta, tokens, done }` object.

// Canonical string values for the two lens types. Reference these (e.g.
// `LensType.JACOBIAN_LENS`) instead of bare string literals to avoid typos.
export const LensType = {
  LOGIT_LENS: 'LOGIT_LENS',
  JACOBIAN_LENS: 'JACOBIAN_LENS',
} as const;
export type LensType = (typeof LensType)[keyof typeof LensType];

export const LENS_TYPES = [LensType.LOGIT_LENS, LensType.JACOBIAN_LENS] as const;

// Lens display mode used by the UI toggle: a single lens type, or DIFF (two
// columns showing each lens's advantage over the other). The single-mode values
// intentionally match `LensType`.
export const LensMode = {
  JACOBIAN_LENS: LensType.JACOBIAN_LENS,
  LOGIT_LENS: LensType.LOGIT_LENS,
  DIFF: 'DIFF',
} as const;
export type LensMode = (typeof LensMode)[keyof typeof LensMode];

export const LENS_MODES = [LensMode.JACOBIAN_LENS, LensMode.LOGIT_LENS, LensMode.DIFF] as const;

// Neuronpedia model id (route segment) for the default jlens model. The
// underlying HF/TransformerLens id (e.g. google/gemma-3-4b-pt) is resolved
// server-side from the model's tlensId, so this must be the slash-free NP id.
export const DEFAULT_JLENS_MODEL_ID = 'qwen3.6-27b';

// Social/OpenGraph preview image for the jlens landing page. Resolved against
// `ASSET_BASE_URL` (site-assets bucket), so this is the path within that bucket.
export const JLENS_METADATA_PATH = '/jlens/jlens.jpg';
export const DEFAULT_LENS_TOP_N = 8;
export const DEFAULT_LENS_TEMPERATURE = 0;
export const DEFAULT_LENS_COMPLETION_TOKENS = 128;
// Ceiling on generated tokens: the chat interface's slider maximum, and the
// bound both `/api/lens/prompt` and `/api/lens/share` validate against. All
// three have to agree — a run the UI lets you generate must be one the prompt
// endpoint accepts and the share endpoint can then store.
export const MAX_LENS_COMPLETION_TOKENS = 2048;
// The single-shot completion interface caps generation lower than chat.
export const MAX_LENS_COMPLETION_TOKENS_COMPLETION = 128;
// Default number of generated tokens for the completion interface (lower than
// the cap above; the chat interface uses DEFAULT_LENS_COMPLETION_TOKENS).
export const DEFAULT_LENS_COMPLETION_TOKENS_COMPLETION = 32;

// Character caps on user-supplied input. Enforced on the frontend (so the UI
// won't let you exceed them) and re-validated on the API (so direct callers are
// rejected too).
//   - Chat: each user message.
//   - Chat: the optional assistant prefill the user types. Enforced on the
//     frontend only — the API can't distinguish a prefill from generated /
//     edited assistant content (which is legitimately longer) since both arrive
//     as assistant messages in the same `chat` payload.
//   - Completion: the single prompt.
export const MAX_LENS_CHAT_USER_CHARS = 1024;
export const MAX_LENS_CHAT_PREFILL_CHARS = 512;
export const MAX_LENS_COMPLETION_PROMPT_CHARS = 1024;

// Steering: default/extent of the strength control (a signed fraction of each
// position's residual norm; negative suppresses the selected readout).
export const DEFAULT_LENS_STEER_STRENGTH = -0.1;
export const MAX_LENS_STEER_STRENGTH = 2;
export const LENS_STEER_STRENGTH_STEP = 0.1;

export interface LensChatMessage {
  role: string;
  content: string;
}

// A single readout to steer on. `token` is the EXACT decoded token string
// (whitespace preserved, e.g. " cat") as it appeared in a read-out slice; the
// server resolves it back to a vocab id. `type` selects which lens's readout
// direction to use (Jacobian: J_bar^T·w_t; Logit: plain unembedding w_t).
export interface LensSteerToken {
  token: string;
  type: LensType;
}

/**
 * The lens request as inference declares it, in camelCase.
 *
 * Derived from the spec rather than hand-written: this used to be a snake_case duplicate of
 * the python model, which meant every field existed in three places and the route had to
 * translate between two of them.
 *
 * Every field is optional because the server supplies defaults; `model` and `type` are the
 * only ones it truly needs, and omitting them surfaces as a 422.
 */
export type LensPromptRequest = Partial<components['schemas']['LensPromptRequest']>;

// The streamed NDJSON frames, mirroring the models in apps/inference's lens/prompt.py.
//
// Hand-written because NDJSON frames never reach openapi.json, so there is nothing to
// generate from. They are not renamed on the way through: inference declares these frames on
// `PublicFrameSchema`, which leaves the field names un-aliased precisely because they are what
// `/api/lens/prompt` publishes and what the stored share blobs contain. Inference's
// test_lens_frame_contract.py pins the names, so a change there fails before it reaches here.

// Lens read-out for one (position, lens_type). All token references are
// decoded STRINGS, never ids.
export interface LensTypeSlice {
  type: LensType;
  // [n_layers][top_n]
  top_tokens: string[][];
  top_probs: number[][];
}

// Per-token chat-span metadata (role / channel / section / message index),
// computed server-side by the engine's tokenize layer (the single source of
// truth for message boundaries). All fields are null on raw-text / reproduction
// requests that carry no chat messages, in which case the client renders the
// tokens plainly. `section` is one of "header" | "content" | "footer" |
// "scaffold"; `channel` is a harmony channel (analysis/final/commentary) or null.
export interface LensTokenSpan {
  message_index?: number | null;
  role?: string | null;
  channel?: string | null;
  section?: string | null;
}

// True on the 2nd..nth position of a character that is split across tokens (an
// emoji, typically). The server repeats the whole glyph in `token` at every
// contributing position so each chip renders it; this flag is what makes that
// repetition reversible, so anything rebuilding TEXT from tokens must go through
// `tokensToText` (jlens-chat-format.ts) rather than joining `token`. Optional
// because runs predating the flag (stored fixtures) simply don't carry it.
export interface LensCharContinuation {
  is_char_continuation?: boolean;
}

// A single chat-formatted prompt token, sent up-front (before inference) so the
// client can render the conversation structure immediately.
export interface LensPromptToken extends LensTokenSpan, LensCharContinuation {
  position: number;
  token: string;
  // Token id, echoed so the client can send it back as `cached_token_ids` on
  // the next turn for prefix-reuse matching.
  id: number;
  is_generated: boolean;
}

// Emitted right after `meta` and before inference begins: the chat-formatted
// prompt tokens (no lens read-outs yet) so the UI can render the full
// conversation shape (user turn + assistant scaffold) right away.
export interface LensPromptTokensMessage {
  kind: 'prompt';
  tokens: LensPromptToken[];
}

// First streamed message: the shared request context.
export interface LensMetaMessage {
  kind: 'meta';
  model: string;
  types: LensType[];
  // Selected layers per lens type (identical for every position).
  layers_by_type: Record<string, number[]>;
  top_n: number;
  prompt_len: number;
  num_completion_tokens: number;
  temperature: number;
  prepend_bos: boolean;
  // Number of leading prompt positions whose read-outs were reused from the
  // client's cache (skipped this run). Token messages are only emitted for
  // positions >= reuse_len; the client keeps its prior results for the rest.
  reuse_len: number;
}

// One per token position: the token plus its per-type lens slices.
export interface LensTokenMessage extends LensTokenSpan, LensCharContinuation {
  kind: 'token';
  position: number;
  token: string;
  // Token id, echoed so the client can send it back as `cached_token_ids` on
  // the next turn for prefix-reuse matching.
  id: number;
  is_generated: boolean;
  results: LensTypeSlice[];
}

// Final streamed message.
export interface LensDoneMessage {
  kind: 'done';
  seq_len: number;
  prompt_len: number;
  vocab_size: number;
  completion: string;
}

export interface LensErrorMessage {
  kind: 'error';
  error: string;
}

export type LensStreamMessage =
  | LensMetaMessage
  | LensPromptTokensMessage
  | LensTokenMessage
  | LensDoneMessage
  | LensErrorMessage;

// Non-streaming response: the same messages buffered into one object.
export interface LensPromptResponse {
  meta: LensMetaMessage;
  tokens: LensTokenMessage[];
  done: LensDoneMessage;
}

export const LENS_TYPE_LABELS: Record<LensType, string> = {
  [LensType.JACOBIAN_LENS]: 'Jacobian Lens',
  [LensType.LOGIT_LENS]: 'Logit Lens',
};

// Preferred order for display columns / default selection.
export const LENS_TYPE_ORDER: LensType[] = [LensType.JACOBIAN_LENS, LensType.LOGIT_LENS];
